wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/InitializeParamsProvider.java

```diff
@@ -0,0 +1,160 @@
+import org.eclipse.lsp4j.SignatureHelpCapabilities;
+import org.eclipse.lsp4j.SymbolCapabilities;
+import org.eclipse.lsp4j.SynchronizationCapabilities;
+import org.eclipse.lsp4j.TextDocumentClientCapabilities;
+import org.eclipse.lsp4j.WorkspaceClientCapabilities;
+import org.eclipse.lsp4j.WorkspaceEditCapabilities;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Utility class that simplifies creation of a language server initialize parameters. */
+@Singleton
+class InitializeParamsProvider {
+  private static final Logger LOG = LoggerFactory.getLogger(InitializeParamsProvider.class);
+
+  private final Registry<Boolean> localityRegistry;
+  private final URI rootUri;
+
+  @Inject
+  InitializeParamsProvider(PathTransformer pathTransformer, RegistryContainer registryContainer) {
+    this.rootUri = pathTransformer.transform(ROOT).toUri();
+    this.localityRegistry = registryContainer.localityRegistry;
+  }
+
+  InitializeParams get(String id) throws LanguageServerException {
+    InitializeParams initializeParams = new InitializeParams();
+    LOG.debug("Initialize params constructing: started");
+
+    Boolean locality = localityRegistry.get(id);
+    LOG.debug("Locality: {}", locality);
+
+    Integer processId = locality ? ProcessIdProvider.get() : null;
+    initializeParams.setProcessId(processId);
+    LOG.debug("Process id: {}", processId);
+
+    String rootPath = Paths.get(rootUri).toAbsolutePath().toString();
+    initializeParams.setRootPath(rootPath);
+    LOG.debug("Root path: {}", rootPath);
+
+    String rootUri = this.rootUri.toString();
+    initializeParams.setRootUri(rootUri);
+    LOG.debug("Root URI: {}", rootUri);
+
+    ClientCapabilities capabilities = ClientCapabilitiesProvider.get();
+    initializeParams.setCapabilities(capabilities);
+    LOG.debug("Client capabilities: {}", capabilities);
+
+    String clientName = ClientCapabilitiesProvider.CLIENT_NAME;
+    initializeParams.setClientName(clientName);
+    LOG.debug("Client name: {}", clientName);
+
+    LOG.debug("Initialize params constructing: finished");
+    return initializeParams;
+  }
+
+  /** Provides the id of current process */
+  private static class ProcessIdProvider {
+    private static Integer processId;
+
+    private static int get() {
+      if (processId == null) {
+        String name = ManagementFactory.getRuntimeMXBean().getName();
+        int prefixEnd = name.indexOf('@');
+        if (prefixEnd != -1) {
+          String prefix = name.substring(0, prefixEnd);
+          try {
+            processId = Integer.parseInt(prefix);
+          } catch (NumberFormatException e) {
+            LOG.error("Failed to recognize the pid of the process", e);
+          }
+        } else {
+          LOG.error("Failed to recognize the pid of the process");
+        }
+      }
+
+      return processId;
+    }
+  }
+
+  /** Provides client capabilities */
+  private static class ClientCapabilitiesProvider {
```

Comment:
Seriously, what's with the providers?

Reply:
That is a very interesting suggestion, however I tend to disagree and let me please explain why. 

InitializeParamsProvider itself is quite big so decomposition here is something that does not require an explanation. However in what way we actually decompose is a very interesting question. In this matter I follow Grady Booch and his book "Object-oriented Analysis and Design" who says that algorithmic decomposition is a necessary part of object-oriented analysis and design, but object-oriented systems start with and emphasize decomposition into classes. In other words I prefer to split the logic into classes, not into fields or variables, but your suggestion is more of a procedural programming.

Reply:
> Yes, but you're not doing algorithmic decomposition, you're just taking a constant and making it more complicated.

That is not true.

Reply:
I correct myself: you're taking a simple static method and making it more complicated. Algorighmic decomposition is decomposition into functions, not classes. I heartily disagree with this style.